File: Section1 - Basics/Recursion/problem5.java
//  Brute Force
class Solution{
    static ArrayList<Long> factorialNumbers(long N){
        // code here
       ArrayList<Long> fact = new ArrayList<Long>();
       
       for(int i=1;i<=N;i++) {
           fact.add(calcFactorial(i));
           if(fact.get(i-1)>N) {
              fact.remove(i-1);
              break;
           }
       }
       return fact;
    }
    
    static long calcFactorial(long n) {
            if(n==0) 
                return 1;
            return n * calcFactorial(n-1);   
    }
    
}

/*Notes
You have to declare method calcFactorial static to call this method from static method factorialNumbers
and return type should also match everywhere
*/

File: Section1 - Basics/Recursion/problem5.js
// Brute force
function calcFactorial(n) {
	if (n == 0) return 1;
	return n * calcFactorial(n - 1);
}

function factorialNumbers(n) {
	let fact = [];
	for (let i = 1; i <= n; i++) {
		fact[i - 1] = calcFactorial(i);
		if (fact[i - 1] > n) {
			fact.pop();
			break;
		}
	}
	return fact;
}

let result = factorialNumbers(6);
console.log(result);

/*Notes
calFactorial calculates factorial of number using recursion in O(n).
fact.pop() method removes last element of array.
*/
